### ISSUE TITLE
Lint for using the same expression for both length and capacity when constructing a vec/string

### ISSUE THREAD

**Issue by sgrif** [rust-lang/rust-clippy#5955]
### What it does

This lint checks for calls to `String::from_raw_parts` and `Vec::from_raw_parts`, and triggers when the same expression is used for both the length and capacity. If folks are interested in a pull request for this lint, I am happy to do the work.

### Categories (optional)

- Kind: pedantic

*What is the advantage of the recommended code over the original code*

When code is written this way, the author likely intended to use a borrowed type instead. The idea for this lint came after noticing a friend of mine consistently reaching for `String` and not `&str`, which requires inventing a capacity. The author will likely use the length in that case. Unless the string/vec was created using `with_capacity` or had `shrink_to_fit` called on it, it's unlikely that the length and capacity are the same.

### Drawbacks

This is trying to catch a beginner mistake, and can potentially trigger on valid code. It's unclear how common this mistake actually is

### Example

```rust
Vec::from_raw_parts(ptr, len, len)
```

Could be written as:

```rust
slice::from_raw_parts(ptr, len)
// or
Vec::from_raw_parts(ptr, len, cap)
```

**Comment by longlb** [comment]
Can I give this one a try?

**Comment by ebroto** [comment]
Oh, I just saw that @sgrif offered themself to implement it. It would be better to hear from them first if they started to work on it already.

**Comment by longlb** [comment]
Not a problem, I'll wait for them to reply. By the way, is there a bug in the repo currently? I just forked it and the build won't compile due to a dependency struct being private.
```
error[E0616]: field kind of struct TyS is private   
--> clippy_lints/src/vec.rs:130:36
    |
130 |     if let ty::Adt(_, substs) = ty.kind {
    |                                    ^^^^ private field
    |
help: a method `kind` also exists, call it with parentheses
    |
130 |     if let ty::Adt(_, substs) = ty.kind() {
    |                                        ^^
```

I have 187 cases of this in my terminal.

**Comment by ebroto** [comment]
Thanks for the heads-up, a fix is in progress in #6007.

This happens when rustc breaks Clippy, we're working on trying to avoid this hassle to contributors :) After that PR passes the tests and is merged, the errors should be gone.

**Comment by km274** [comment]
Hi 👋🏻, I'm an experienced software engineer but am new to Rust. I picked this as a first issue since it seems like it doesn't exactly need to be done in a rush 😋. I'll get to work in my spare time.
